pkg/client/fake/client.go

```diff
@@ -134,15 +154,33 @@ func (c *fakeClient) Status() client.StatusWriter {
 	return &fakeStatusWriter{client: c}
 }
 
-func getGVRFromObject(obj runtime.Object) (schema.GroupVersionResource, error) {
-	gvk, err := apiutil.GVKForObject(obj, scheme.Scheme)
+func getGVRFromObject(obj runtime.Object, scheme *runtime.Scheme) (schema.GroupVersionResource, error) {
+	gvk, err := apiutil.GVKForObject(obj, scheme)
 	if err != nil {
 		return schema.GroupVersionResource{}, err
 	}
 	gvr, _ := meta.UnsafeGuessKindToResource(gvk)
 	return gvr, nil
 }
 
+func getGVKFromList(list runtime.Object, scheme *runtime.Scheme) (schema.GroupVersionKind, error) {
+	gvk, err := apiutil.GVKForObject(list, scheme)
+	if err != nil {
+		return schema.GroupVersionKind{}, err
+	}
+
+	if gvk.Kind == "List" {
+		return schema.GroupVersionKind{}, fmt.Errorf("cannot derive GVK for generic List type %T (kind %q)", list, gvk)
```

Comment:
This check is necessary to detect the `corev1.List` and `metav1.List` type, which don't conform to the `SingularTypeList` format expected here. Without this check the returned GVK has an empty `Kind`.

It seems worthwhile to consolidate this derivation into a single helper function, since it's also used in varying forms in these places:
https://github.com/kubernetes-sigs/controller-runtime/blob/2ef903b2a5ed4cf8adde14a27994a291388a1789/pkg/client/client_cache.go#L60
https://github.com/kubernetes-sigs/controller-runtime/blob/df7c11e379403b7377ec52e54fd1baf587d3ac29/pkg/client/unstructured_client.go#L115
https://github.com/kubernetes-sigs/controller-runtime/blob/df7c11e379403b7377ec52e54fd1baf587d3ac29/pkg/cache/informer_cache.go#L67